backend/tessend/tessend.go

```diff
@@ -133,42 +154,80 @@ func (t *Tessend) handleWatchEvent(watchEvent store.WatchEventTessenConfig) {
 	t.interrupt <- tessen
 }
 
-// start starts a new scheduler for tessen
-func (t *Tessend) start(tessen *corev2.TessenConfig) error {
-	// Guard against updates while the daemon is shutting down
-	if err := t.ctx.Err(); err != nil {
-		return err
+// startRingUpdates starts a loop to periodically update the ring.
+func (t *Tessend) startRingUpdates() {
+	ticker := time.NewTicker(ringUpdateInterval)
+	defer ticker.Stop()
+	t.updateRing()
+	for {
+		select {
+		case <-t.ctx.Done():
+			return
+		case <-ticker.C:
+			t.updateRing()
+		}
 	}
+}
 
-	go t.schedule(tessen)
+// updateRing adds/updates the ring with a given key.
+func (t *Tessend) updateRing() {
+	if err := t.ring.Add(t.ctx, t.backendID, ringBackendKeepalive); err != nil {
+		logger.WithField("key", t.backendID).WithError(err).Error("error adding key to the ring")
+	} else {
+		logger.WithField("key", t.backendID).Debug("added a key to the ring")
+	}
+}
 
-	return nil
+// watchRing watches the ring and handles ring events.
+func (t *Tessend) watchRing(ctx context.Context, tessen *corev2.TessenConfig) {
+	wc := t.ring.Watch(ctx, "tessen", 1, int(t.interval), "")
+	go t.handleEvents(tessen, wc)
 }
 
-func (t *Tessend) schedule(tessen *corev2.TessenConfig) {
-	timer := time.NewTimer(time.Duration(time.Second * time.Duration(t.interval)))
-	defer timer.Stop()
+// handleEvents logs different ring events and triggers tessen to run if applicable.
+func (t *Tessend) handleEvents(tessen *corev2.TessenConfig, ch <-chan ringv2.Event) {
+	for event := range ch {
+		switch event.Type {
+		case ringv2.EventError:
+			logger.WithError(event.Err).Error("ring event error")
+		case ringv2.EventAdd:
+			logger.WithField("values", event.Values).Debug("ring event add")
+		case ringv2.EventRemove:
+			logger.WithField("values", event.Values).Debug("ring event remove")
+		case ringv2.EventTrigger:
+			logger.WithField("values", event.Values).Debug("ring event trigger")
+			// only trigger tessen if the next backend in the ring is this backend
+			if event.Values[0] == t.backendID {
+				if t.enabled(tessen) {
+					t.collectAndSend()
+				}
+			}
+		case ringv2.EventClosing:
+			logger.Debug("ring event closing")
+		}
+	}
+}
+
+// start starts the tessen service.
+func (t *Tessend) start(tessen *corev2.TessenConfig) {
 	// Attempt to send data immediately if tessen is enabled
 	if t.enabled(tessen) {
 		t.collectAndSend()
 	}
 
+	ctx, cancel := context.WithCancel(t.ctx)
+	t.watchRing(ctx, tessen)
+
 	for {
 		select {
 		case <-t.ctx.Done():
+			cancel()
 			return
 		case config := <-t.interrupt:
-			defer func() {
-				go t.schedule(config)
-			}()
-			return
-		case <-timer.C:
-		}
-		// Attempt to send data if tessen is enabled
-		if t.enabled(tessen) {
-			t.collectAndSend()
+			cancel()
+			ctx, cancel = context.WithCancel(t.ctx)
+			t.watchRing(ctx, config)
```

Comment:
We need t.watchRing to be called when t.interval updates. I'd suggest sending an interrupt when that change is detected.